### PR TITLE
Fix deg to rad conversion in L1TkMuonProducer ("ManTra" version) for HLT TDR [11_1_X]

### PR DESCRIPTION
**Pull request by dildick**
Backport of https://github.com/cms-sw/cmssw/pull/31341

FYI @l-cadamuro @rekovic @BenjaminRS

### CODE REVIEW

**Comment by dpiparo** [comment]
Is there a reason why not to use the existing routine (https://cmssdt.cern.ch/lxr/source/DataFormats/Math/interface/angle_units.h)?

**Comment by silviodonato** [comment]
@dildick please sync this PR with https://github.com/cms-sw/cmssw/pull/31341

**Comment by dildick** [comment]
@silviodonato Per your comment here https://github.com/cms-sw/cmssw/pull/31341#issuecomment-686576626, I'm not planning to backport the clean-up to 11_1_X, just the essential fix.

**Comment by silviodonato** [comment]
@dildick In https://github.com/cms-sw/cmssw/pull/31341#issuecomment-686576626 I supposed we merge quickly the "essential" bugfix and we get rid of `deg_to_rad` later. Given that https://github.com/cms-sw/cmssw/pull/31341 is already merged, I would prefer to have #31341 also in `11_1_X`. It is better to have the code synchronized in the two releases because of 1) possible further backports from `11_2_X` to `11_1_X` are simpler (no code difference) 2) in case there is a bug in `11_1_X` we can find it also in `11_2_X`, eg. with the release validation.

**Comment by davidlange6** [comment]
How do you manage production inconsistencies if pushing this back to 11_1?

On Sep 8, 2020, at 7:18 PM, Silvio Donato <notifications@github.com<mailto:notifications@github.com>> wrote:



@dildick<https://github.com/dildick> In #31341 (comment)<https://github.com/cms-sw/cmssw/pull/31341#issuecomment-686576626> I supposed we merge quickly the "essential" bugfix and we get rid of deg_to_rad later. Given that #31341<https://github.com/cms-sw/cmssw/pull/31341> is already merged, I would prefer to have #31341<https://github.com/cms-sw/cmssw/pull/31341> also in 11_1_X. It is better to have the code synchronized in the two releases because of 1) possible further backports from 11_2_X to 11_1_X are simpler (no code difference) 2) in case there is a bug in 11_1_X we can find it also in 11_2_X, eg. with the release validation.

**Comment by silviodonato** [comment]
> How do you manage production inconsistencies if pushing this back to 11_1? 

My understanding is that the HLT production is done and if they want to re-run the L1 trigger in the future they want to have this bugfix (the conversion `deg_to_rad` is totally wrong right now).

@rappoccio @trtomei @fwyzard @dildick @rekovic do you need this backport at all? If I remember correctly, the plan is to re-run the l1tkmuon on-fly during the re-HLT (ie. in 11_2_X).

**Comment by davidlange6** [comment]
Wrong but consistent versus sometimes right and sometimes wrong.. Clearly if production is going to change and will pick up bug fixes and redo things, then this ought to go in, but until then consistency is probably better

On Sep 9, 2020, at 10:54 AM, Silvio Donato <notifications@github.com<mailto:notifications@github.com>> wrote:



How do you manage production inconsistencies if pushing this back to 11_1?

My understanding is that the HLT production is done and if they want to re-run the L1 trigger in the future they want to have this bugfix (the conversion deg_to_rad is totally wrong right now).

@rappoccio<https://github.com/rappoccio> @trtomei<https://github.com/trtomei> @fwyzard<https://github.com/fwyzard> @dildick<https://github.com/dildick> @rekovic<https://github.com/rekovic> do you need this backport at all? If I remember correctly, the plan is to re-run the l1tkmuon on-fly during the re-HLT (ie. in 11_2_X).

**Comment by fwyzard** [comment]
> do you need this backport at all?

Yes.

> If I remember correctly, the plan is to re-run the l1tkmuon on-fly during the re-HLT (ie. in 11_2_X).

11.2.x is not compatible with the 11.0.x/11.1.x RAW/DIGIs, so all future HLT processing will be done with 11.1.x.
So, I would like the 11.1.x release series to includes all available bug fixes.

**Comment by dildick** [comment]
@fwyzard This commit https://github.com/cms-sw/cmssw/pull/31342/commits/f03291805a6a3cb1e4d3cf8729c18db96540c038 is the bugfix for 11_1_X. This commit https://github.com/cms-sw/cmssw/pull/31341/commits/ab1423a4bbb1dc7a44a3b48b31662dbad9ef296b (only in 11_2_X) is a clean-up of the `deg_to_rad` and `to_mpi_pi` functions.

**Comment by fwyzard** [comment]
11.1.x needs to get all the bugfixes for running (all or part of) the L1T from DIGIs/RAW.

My opinion is that doing a partial backport is more work and more error prone than ding a full backport (i.e. just cherry-pick the same changes for both branches).
If people think that some "clean up" doesn't need to be back ported, I won't insist on that.

**Comment by fwyzard** [comment]
@davidlange6 
> How do you manage production inconsistencies if pushing this back to 11_1?

> Wrong but consistent versus sometimes right and sometimes wrong.. Clearly if production is going to change and will pick up bug fixes and redo things, then this ought to go in, but until then consistency is probably better

Consistency come from using the same release.

Moving to a more recent release (both for production and for people running recipes) to get the bugfixes that were included in the meantime *is* the reason for making new 11.1.x releases.

**Comment by davidlange6** [comment]
I’d call it “a” reason for making new 11.1.x releases. There are also potential changes to random technical packages (or eg generators) that may be needed for some samples. These can conflict if consistency between samples used for an analysis is important and the last release digit has been used for changes like this one (One can clearly decide that consistency is not important) Clearly not being able to move to 11.2.x is an important consideration. My question is was how this would be handled not that it shouldn’t be done..

On Sep 9, 2020, at 3:43 PM, Andrea Bocci <notifications@github.com<mailto:notifications@github.com>> wrote:



@davidlange6<https://github.com/davidlange6>

How do you manage production inconsistencies if pushing this back to 11_1?

Wrong but consistent versus sometimes right and sometimes wrong.. Clearly if production is going to change and will pick up bug fixes and redo things, then this ought to go in, but until then consistency is probably better

Consistency come from using the same release.

Moving to a more recent release (both for production and for people running recipes) to get the bugfixes that were included in the meantime is the reason for making new 11.1.x releases.

**Comment by fwyzard** [comment]
> I’d call it “a” reason for making new 11.1.x releases.

Of course - let me rephrase: it's a good enough reason to request a new 11.1.x release.

**Comment by dildick** [comment]
Commit https://github.com/cms-sw/cmssw/pull/31342/commits/41dea91284cda94309422e56391572d42aacc658 is the cherry-picked version of https://github.com/cms-sw/cmssw/commit/ab1423a4bbb1dc7a44a3b48b31662dbad9ef296b in 11_2_X.

**Comment by dildick** [comment]
I checked the performance with the clean-up. 
[Eff_L1TrackMuon_pt_Pt20_.pdf](https://github.com/cms-sw/cmssw/files/5196285/Eff_L1TrackMuon_pt_Pt20_.pdf)
[Eff_L1TrackMuon_eta_Pt20_.pdf](https://github.com/cms-sw/cmssw/files/5196283/Eff_L1TrackMuon_eta_Pt20_.pdf)
[Eff_L1TrackMuon_phi_Pt20_.pdf](https://github.com/cms-sw/cmssw/files/5196284/Eff_L1TrackMuon_phi_Pt20_.pdf)

They look just like the ones without the clean-up
This seems to recover (most of)  the inefficiency. 
- pt: https://its.cern.ch/jira/secure/attachment/249239/Screen%20Shot%202020-09-02%20at%209.36.18%20PM.png
- abs(eta): https://its.cern.ch/jira/secure/attachment/249240/Screen%20Shot%202020-09-02%20at%209.36.56%20PM.png
- phi: https://its.cern.ch/jira/secure/attachment/249241/Screen%20Shot%202020-09-02%20at%209.37.29%20PM.png

The drop at |eta|~1.6 is under investigation by @l-cadamuro.

**Comment by silviodonato** [comment]
please test

**Comment by silviodonato** [comment]
> I checked the performance with the clean-up.
> [Eff_L1TrackMuon_pt_Pt20_.pdf](https://github.com/cms-sw/cmssw/files/5196285/Eff_L1TrackMuon_pt_Pt20_.pdf)
> [Eff_L1TrackMuon_eta_Pt20_.pdf](https://github.com/cms-sw/cmssw/files/5196283/Eff_L1TrackMuon_eta_Pt20_.pdf)
> [Eff_L1TrackMuon_phi_Pt20_.pdf](https://github.com/cms-sw/cmssw/files/5196284/Eff_L1TrackMuon_phi_Pt20_.pdf)
> 
> They look just like the ones without the clean-up:
> 
>     * pt: https://its.cern.ch/jira/secure/attachment/249239/Screen%20Shot%202020-09-02%20at%209.36.18%20PM.png
> 
>     * abs(eta): https://its.cern.ch/jira/secure/attachment/249240/Screen%20Shot%202020-09-02%20at%209.36.56%20PM.png
> 
>     * phi: https://its.cern.ch/jira/secure/attachment/249241/Screen%20Shot%202020-09-02%20at%209.37.29%20PM.png
> 
> 
> The drop at |eta|~1.6 is under investigation by @l-cadamuro.

do you get the drop at |eta|~1.6 even in `11_2_X`?
It looks like a possible further bug, I would merge this bugfix meanwhile.

**Comment by dildick** [comment]
Hi Silvio, the drop at 1.6 is also there in 11_2_X.

**Comment by kpedro88** [comment]
+upgrade

**Comment by rekovic** [comment]
+1

**Comment by qliphy** [comment]
+1
